### ISSUE TITLE
Coordination: last items before branching 2.0.x

### ISSUE THREAD

**Issue by rgommers** [numpy/numpy#25918]
Posting as a new issue rather than trying to coordinate with comments on various other PRs. In the community meeting earlier this week we discussed that we are almost ready, and @charris planned to branch `maintenance/2.0.x` today. We still have a few more things to be done though it looks like. 

Here is what I'd propose we do today:

- [x] Merge gh-25897
- [x] Finalize and merge gh-25911 (@rgommers)
- [x] Revert gh-25168 (because of gh-25916) (Done in #25922)
- [x] Then build new nightlies (@rgommers) (Triggered manually by @mattip)
- [x] Once the nightlies are up: ensure things work with using `npy_2_compat.h` in SciPy (https://github.com/scipy/scipy/pull/20108), getting that merged and uploading new SciPy nightlies

After that (tomorrow?):
- [x] Fix up gh-25168 and get it merged again (@mtsokol and @ngoldbaum)
- [x] If blocking for 2.0.0rc1: get gh-25802 in (@seberg)
- [x] Last ABI-breaking PR, ~still to be opened by @seberg~: gh-25943

Other things:
- [x] Another pass over the release notes (@rgommers) (gh-25921, gh-25937)
- [x] May want to get gh-25914 in if there are no objections to that one

Everything else under the [2.0.0 milestone](https://github.com/numpy/numpy/milestone/118) seems like it can wait.

**Comment by rgommers** [comment]
I missed a comment from @ngoldbaum elsewhere, so I'll wait for at least his input here before touching the `copy` keyword issue.

**Comment by ngoldbaum** [comment]
I’m going to try to work on a PR to fix the issues with the `copy` keyword this afternoon. If I can’t figure it out I’ll send in a revert.

**Comment by rgommers** [comment]
@charris I'm having a go at release notes updates now.

**Comment by charris** [comment]
@rgommers Great.

**Comment by mattip** [comment]
Wheel building failed, see #25923

**Comment by seberg** [comment]
There is one more PR unless it wrecks havoc.  The DType ones were a bit stalled, there is a follow up to bump the itemsize to `intp` and finalize the dtype flags changes.
(I.e. the actual ABI break, I was quite careful to prepare the break in the first PR, but the back and forth delayed the follow up.)

**Comment by rgommers** [comment]
Thanks @seberg. Let me ask a few questions to really spell it out, because Chuck was already planning to call it good and create the release branch once the copy keyword regression was fixed. 

1. You still want to get exactly two things in before 2.0.0rc1: gh-25802 and a to-be-opened PR (which I think addresses the second item of gh-25820)?
2. Do those two depend on each other?
3. How much time do you estimate you need?
4. Is it necessary to do another round of "build nightlies, then build nightlies of scipy straight after, then test scikit-learn/astropy"? (probably yes?) If so, can we do it once both PRs are in (since shepherding all the nightly builds is pretty time-consuming)?

**Comment by seberg** [comment]
It shouldn't take long to create the follow up, but right now the most time was always waiting for review (I will skip renaming `elsize` to `itemsize` for now to make the change non-huge).

The annoyance is fixing `pybind11` and SciPy needs that due to a single line of code:
```
scipy/spatial/src/distance_pybind.cpp:    desc.element_size = arr.itemsize();
```
I can even do that, but not quite sure how long review would take.

**Comment by rgommers** [comment]
Any chance we can work around that in SciPy? Getting `pybind11` to not only fix/review something but do a release that SciPy can use may be difficult.

**Comment by seberg** [comment]
> Any chance we can work around that in SciPy?

Seems that while `pybind11` avoids importing NumPy explicitly, SciPy does not in the module (that I noticed is bad) using pybind11, so using the NumPy provided macros should be trivial, yes.
(I am confused why pybind11 avoids it so much considering SciPy does it...)

**Comment by rgommers** [comment]
Yes, that looks expected @jakevdp. For notes on the changes, see https://github.com/numpy/numpy/blob/54469e2e21be0bb5059e71f91542fd0ac8516d82/doc/source/release/2.0.0-notes.rst#L1217-L1242 

This SciPy PR to adapt to the changes may be useful too: https://github.com/scipy/scipy/pull/20172

**Comment by seberg** [comment]
The thing that I thought was in the release notes/migration guide: Avoid using `array` if you have `copy=False` since  `asarray` is the clearer version.  (A long time ago, that had much more overhead, which is probably the main reason why code has it.)

**Comment by rgommers** [comment]
> The thing that I thought was in the release notes/migration guide: Avoid using `array` if you have `copy=False` since `asarray` is the clearer version

Good point. It's missing from the migration guide, which is the right place I think. I'll take it along in my next PR for release notes updates.

**Comment by jakevdp** [comment]
Thanks!

**Comment by rgommers** [comment]
Looks like the last PRs we need are up!

One thing I'd like to suggest is to do a `2.0.0b1` release, which is likely but not guaranteed to have a stable C API and ABI. Doing an `rc1` which has a hard guarantee to be stable within days of merging large breaking changes is probably a bit too risky, so it'd be nice to have a 1-2 week escape hatch.

I imagine the following rollout scenario for releases to PyPI:

1. numpy `2.0.0b1`
2. scipy/pandas/matplotlib/etc. make RC1 releases
3. announcement that `pip install <your-deps> --pre` works from PyPI, give a 3-7 day window to catch major issues
4. numpy `2.0.0rc1` (guaranteed stable now)
5. scipy/pandas/matplotlib/etc. do a final release, compatible with 2.0 and 1.xx
6. numpy `2.0.0rc2` (surely we'll need it for something)
7. numpy `2.0.0`

**Comment by seberg** [comment]
I am happy with a `b1` to not signal ABI stability right away.  I don't particularly expect ABI changes (as opposed to API ones; the main place where I see a meaningful ABI break would be the new DType API and nobody will release a new DType in the next month).

**Comment by charris** [comment]
My impression is that error reports are either immediate, or show up after the first week, so things might move a bit slower than planned. But that is OK as long as something is out there.

We can drop Python 3.9 support in main after the branch, I suggest moving more tests to Python 3.12 where they can hang around for a couple of years.

**Comment by charris** [comment]
Looks like we are almost ready.

**Comment by rgommers** [comment]
Yep, I think once the last release notes PR is merged, the branch can be created.

**Comment by seberg** [comment]
You were right, seems if matplotlib is fine (enough).  Doc builds are working again and the website is up to date again.

**Comment by rgommers** [comment]
@seberg are you still planning to make a change in between 2.0.0b1 and 2.0.0rc1 that turns the warning for `pybind11` usage of private API into a hard error? Or, since it kinda works for now, do it only in `main` after `pybind11` is fixed up and released?

**Comment by seberg** [comment]
:shrug: that huge warning is seems only semi-useful, it would be nice to make it a hard error by the time we release, otherwise we don't protect users from potentially bad results/crashes due to a module that wasn't recompiled.
I don't care much about `rc1` specifically it just seems like a plausible/reasonable warning shot for pybind11.

**Comment by rgommers** [comment]
Well, it must be done either before `rc1` or not at all. The way it looks now, it actually seems fairly safe to disable the warning, leaving the compat shim in place for 2.0, and make it a hard error on `main` in a month or so.

Probably best to wait and see how a `pybind11` release is coming along, and decide in a week or so?

**Comment by rgommers** [comment]
The release branch is created and `2.0.0b1` is up on PyPI, so I will close this issue. There are a few things we need to do before `2.0.0rc1`, including ones touched upon in this issue, so I will create a new tracking issue for that.

Thanks all!
